### PR TITLE
2/2 - Commonwell Document Retrieval

### PR DESCRIPTION
**Pull request by leite08**
Ref #227

### Dependencies

- Upstream: https://github.com/metriport/metriport/pull/89
 - Downstream: none

### Description

Add document retrieval/download from Commonwell.

Downloads files on the local file system. Ideally, we would store those on S3 at some point, maybe when we move that code to the application?

### Release Plan

- anytime

### CODE REVIEW

File: packages/packages/commonwell-cert-runner/src/document-consumption.ts [hunk 1]
```diff
@@ -112,15 +106,27 @@ export async function documentConsumption(commonWell: CommonWell, queryMeta: Req
...
+
+export async function retrieveDocument(
+  commonWell: CommonWell,
+  queryMeta: RequestMetadata,
+  doc: Document
+): Promise<void> {
+  // E2: Document Retrieve
+  console.log(`>>> E2c: Retrieve documents using FHIR (REST)`);
+
+  const fileName = `./commonwell_${nanoid()}.file`;
```
Comment (leite08) [inline packages/packages/commonwell-cert-runner/src/document-consumption.ts:124]: Using `nanoid()` bc `id` and `identifier.value` return a string with `:`, and could potentially return invalid chars for filename.

We should store the file name on the DB at some point, so we can have any alphanumeric file names.

File: packages/packages/commonwell-cert-runner/src/document-consumption.ts [hunk 2]
```diff
@@ -112,15 +106,27 @@ export async function documentConsumption(commonWell: CommonWell, queryMeta: Req
...
+  const outputStream = fs.createWriteStream(fileName, {
+    // the default is UTF-8, avoid changing the encoding if we don't know the file we're downloading
+    encoding: null,
+  });
```
Comment (leite08) [inline packages/packages/commonwell-cert-runner/src/document-consumption.ts:128]: Ideally we would pipe the stream from the download to a S3 upload, maybe when we take this to the app?

For inspiration: https://stackoverflow.com/questions/37336050/pipe-a-stream-to-s3-upload

File: packages/packages/commonwell-cert-runner/src/payloads.ts [hunk 3]
```diff
@@ -5,16 +5,16 @@ import {
...
-import { getEnvOrFail } from "./util";
+import { adjectives, animals, colors, uniqueNamesGenerator } from "unique-names-generator";
+import { getCertificateContent, getEnvOrFail } from "./util";
 
 const commonwellOID = getEnvOrFail("COMMONWELL_OID");
-const commonwellCertificateContent = getEnvOrFail("COMMONWELL_CERTIFICATE_CONTENT");
-const commonwellCertificate = getEnvOrFail("COMMONWELL_CERTIFICATE");
 const commonwellOrgName = getEnvOrFail("COMMONWELL_ORG_NAME");
+const commonwellCertificate = getEnvOrFail("COMMONWELL_CERTIFICATE");
+const commonwellCertificateContent = getCertificateContent(commonwellCertificate);
```
Comment (leite08) [inline packages/packages/commonwell-cert-runner/src/payloads.ts:17]: @Orta21

File: packages/packages/commonwell-cert-runner/src/document-consumption.ts (hunk 2 above)
Comment (Goncharo) [reply packages/packages/commonwell-cert-runner/src/document-consumption.ts:128]: We'll definitely need to think about storage for these files after adding to the api. Ultimately, we'll want to process these into FHIR

File: packages/packages/commonwell-sdk/src/client/commonwell.ts [hunk 4]
```diff
@@ -661,6 +665,29 @@ export class CommonWell {
     return documentQueryResponseSchema.parse(res.data);
   }
 
+  /**
+   * Retrieves a Document and store it on the local file system.
```
Comment (Goncharo) [inline packages/packages/commonwell-sdk/src/client/commonwell.ts:669]: Is this true? Can't the WriteStream be to something else other than the local fs?

File: packages/packages/commonwell-sdk/src/client/commonwell.ts [hunk 5]
```diff
@@ -646,13 +649,14 @@ export class CommonWell {
...
+   * @see {@link https://specification.commonwellalliance.org/services/data-broker/cha-broker-api-reference#104-document-query|Use case}
+   * @see {@link https://specification.commonwellalliance.org/services/data-broker/protocol-operations-data-broker#8781-find-documents|API spec}
    *
    * @param meta       Metadata about the request.
    * @param subjectId  The ID as defined by the specification: [system]|[code] where 'system' is the OID
    * of the organization and 'code' is the first (numeric) part of the patient ID.
-   * @returns {DocumentQueryResponse}
+   * @returns {Promise<DocumentQueryResponse>}
    */
   async queryDocument(meta: RequestMetadata, subjectId: string): Promise<DocumentQueryResponse> {
```
Comment (Goncharo) [inline packages/packages/commonwell-sdk/src/client/commonwell.ts:661]: `queryDocuments`?

**Review by Goncharo — CHANGES_REQUESTED** [review]

**Review by Orta21 — APPROVED** [review]

File: packages/packages/commonwell-sdk/src/client/commonwell.ts (hunk 4 above)
Comment (leite08) [reply packages/packages/commonwell-sdk/src/client/commonwell.ts:669]: Yep, outdated comment. Good catch!

File: packages/packages/commonwell-sdk/src/client/commonwell.ts (hunk 5 above)
Comment (leite08) [reply packages/packages/commonwell-sdk/src/client/commonwell.ts:661]: Addressed on the upstream PR

**Review by Goncharo — APPROVED** [review]
🚢

**Pull request merged**